apiserver/billing/odoostorage/odoo/odoo8/odoo8.go

```diff
@@ -122,11 +160,104 @@ func (s *oodo8Storage) List(ctx context.Context) ([]billingv1.BillingEntity, err
 }
 
 func (s *oodo8Storage) Create(ctx context.Context, be *billingv1.BillingEntity) error {
-	return errors.New("not implemented")
+	l := klog.FromContext(ctx)
+
+	if be == nil {
+		return errors.New("billing entity is nil")
+	}
+	company, accounting, err := mapBillingEntityToPartners(*be, s.countryIDs)
+	if err != nil {
+		return fmt.Errorf("failed mapping billing entity to partners: %w", err)
+	}
+
+	inflight := uuid.New().String()
+	l = l.WithValues("debug_inflight", inflight)
+	company.Inflight = model.NewNullable(inflight)
+	accounting.Inflight = model.NewNullable(inflight)
+	setStaticCompanyFields(&company)
+	setStaticAccountingContactFields(&accounting)
+
+	session, err := s.sessionCreator(ctx)
+	if err != nil {
+		return err
+	}
+	o := model.NewOdoo(session)
+
+	companyID, err := o.CreatePartner(ctx, company)
+	if err != nil {
+		return fmt.Errorf("error creating company: %w", err)
+	}
+	l.Info("created company (parent)", "id", companyID)
+
+	accounting.Parent = model.OdooCompositeID{ID: companyID, Valid: true}
+	accountingID, err := o.CreatePartner(ctx, accounting)
+	if err != nil {
+		return fmt.Errorf("error creating accounting contact: %w", err)
+	}
+	l.Info("created accounting contact", "id", accountingID, "parent_id", companyID)
+
+	// reset inflight flag
+	if err := o.UpdateRawPartner(ctx, []int{companyID, accountingID}, map[string]any{
+		"x_control_api_inflight": false,
+	}); err != nil {
+		return fmt.Errorf("error resetting inflight flag: %w", err)
+	}
+
+	nbe, err := s.Get(ctx, odooIDToK8sID(accountingID))
+	if err != nil {
+		return fmt.Errorf("error fetching newly created billing entity: %w", err)
+	}
+	*be = *nbe
+	return nil
 }
 
 func (s *oodo8Storage) Update(ctx context.Context, be *billingv1.BillingEntity) error {
-	return errors.New("not implemented")
+	l := klog.FromContext(ctx)
+
+	if be == nil {
+		return errors.New("billing entity is nil")
+	}
+
+	company, accounting, err := mapBillingEntityToPartners(*be, s.countryIDs)
+	if err != nil {
+		return fmt.Errorf("failed mapping billing entity to partners: %w", err)
+	}
+
+	origCompany, origAccounting, err := s.get(ctx, be.Name)
+	if err != nil {
+		return fmt.Errorf("error fetching billing entity to update: %w", err)
+	}
+
+	session, err := s.sessionCreator(ctx)
+	if err != nil {
+		return err
+	}
+	o := model.NewOdoo(session)
+
+	fco, err := filterFields(company, companyUpdateAllowedFields)
```

Comment:
Shouldn't trying to edit a field that you're not allowed to edit raise an error? Other resources have similar restrictions, e.g. editing a PVC's storageClassName returns an error along the lines of "# * spec: Forbidden: spec is immutable after creation except resources.requests for bound claims" followed by the diff of the failed edit)

Reply:
Maybe the name is not quite right.

There's a ton of fields we don't want to override in Odoo. Sadly Odoo overrides them with an empty value if the key in JSON is present even if the value is `null` or `false`.

The only chance to not override them is removing the key from the serialised object. `json:"blub,omitempty"` can't do that since we use custom marshalling and unmarshalling everywhere.

Reply:
I improved the comment in the code.